### PR TITLE
Fix handling of FFI arguments

### PR DESCRIPTION
**Pull request by nagisa**
r? @eddyb @nikomatsakis or whoever else.

cc @alexcrichton @rust-lang/core

The strategy employed here was to essentially change code we generate from

``` llvm
  %s = alloca %S ; potentially smaller than argument, but never larger
  %1 = bitcast %S* %s to { i64, i64 }*
  store { i64, i64 } %0, { i64, i64 }* %1, align 4
```

to 

``` llvm
  %1 = alloca { i64, i64 } ; the copy of argument itself
  store { i64, i64 } %0, { i64, i64 }* %1, align 4
  %s = bitcast { i64, i64 }* %1 to %S* ; potentially truncate by casting to a pointer of smaller type.
```

### CODE REVIEW

**Comment by nagisa** [comment]
Hmm, it clashes with debuginfo.

File: src/librustc_trans/base.rs [hunk 1]
```diff
@@ -1663,21 +1663,30 @@ impl<'blk, 'tcx> FunctionContext<'blk, 'tcx> {
...
+                        let lltemp = alloca(bcx, llarg_ty, "");
+                        let b = &bcx.build();
+                        arg.store_fn_arg(b, &mut llarg_idx, lltemp);
+                        // And coerce the temporary into the type we expect.
+                        b.pointercast(lltemp, arg.memory_ty(bcx.ccx()).ptr_to())
+                    };
+
+                    // FIXME: hacky lol?
+                    datum::Datum::new(lltmp, arg_ty,
+                                      datum::Lvalue::new("datum::lvalue_scratch_datum"))
```
Comment (eddyb) [inline src/librustc_trans/base.rs]: This won't work with `debuginfo`, which wants an `alloca` with no casts on top.
I would suggest giving it the `alloca` before the cast and hope that it doesn't care about the type (debuginfo metadata should describe the actual value in the `alloca` all by itself).
cc @michaelwoerister

**Comment by alexcrichton** [comment]
Thanks for the quick fix @nagisa!

**Comment by nagisa** [comment]
Added a check for debuginfo thing. All should work now.

If there’s any hard-pressing nits, feel to fork and make another PR as I’m likely to be in bed for a while now.

**Comment by eddyb** [comment]
Miscompiling the compiler, resulting in this for stage2 `libcore`:
```
thread 'rustc' panicked at 'RefCell<T> already borrowed', ../src/libcore/cell.rs:492
```
@alexcrichton Serious question: why is `RUST_BACKTRACE=1` not set on the bots? I assume it has to do with some tests not taking that very well?

**Comment by alexcrichton** [comment]
No particular reason, it just hasn't been done yet. We'd probably benefit from just setting it in the build system directly.

**Comment by sanxiyn** [comment]
Another case of #33844.

**Comment by alexcrichton** [comment]
Discussed on IRC and accepted for a beta backport

**Pull request merged**